packages/ozone/src/mod-service/index.ts

```diff
@@ -0,0 +1,637 @@
+    // Ideally, this should never happen because the caller of this method should only call this
+    // after ensuring that the suspended or muted subjects are being reversed
+    if (!subject.reverseMute && !subject.reverseSuspend) {
+      return null
+    }
+
+    let builder = this.db.db
+      .selectFrom('moderation_event')
+      .where('subjectDid', '=', subject.subject.did)
+
+    if (subject.subject.recordPath) {
+      builder = builder.where(
+        'subjectUri',
+        'like',
+        `%${subject.subject.recordPath}%`,
+      )
+    }
+
+    // Means the subject was suspended and needs to be unsuspended
+    if (subject.reverseSuspend) {
+      builder = builder
+        .where('action', '=', 'com.atproto.admin.defs#modEventTakedown')
+        .where('durationInHours', 'is not', null)
+    }
+    if (subject.reverseMute) {
+      builder = builder
+        .where('action', '=', 'com.atproto.admin.defs#modEventMute')
+        .where('durationInHours', 'is not', null)
+    }
+
+    return await builder
+      .orderBy('id', 'desc')
+      .selectAll()
+      .limit(1)
+      .executeTakeFirst()
+  }
+
+  async getSubjectsDueForReversal(): Promise<ReversalSubject[]> {
+    const now = new Date().toISOString()
+    const subjects = await this.db.db
+      .selectFrom('moderation_subject_status')
+      .where('suspendUntil', '<', now)
+      .orWhere('muteUntil', '<', now)
+      .selectAll()
+      .execute()
+
+    return subjects.map((row) => ({
+      subject: subjectFromStatusRow(row),
+      reverseSuspend: !!row.suspendUntil && row.suspendUntil < now,
+      reverseMute: !!row.muteUntil && row.muteUntil < now,
+    }))
+  }
+
+  async isSubjectSuspended(did: string): Promise<boolean> {
+    const res = await this.db.db
+      .selectFrom('moderation_subject_status')
+      .where('did', '=', did)
+      .where('recordPath', '=', '')
+      .where('suspendUntil', '>', new Date().toISOString())
+      .select('did')
+      .limit(1)
+      .executeTakeFirst()
+    return !!res
+  }
+
+  async revertState({
+    createdBy,
+    createdAt,
+    comment,
+    action,
+    subject,
+  }: ReversibleModerationEvent): Promise<ModerationEventRow> {
+    const isRevertingTakedown =
+      action === 'com.atproto.admin.defs#modEventTakedown'
+    this.db.assertTransaction()
+    const result = await this.logEvent({
+      event: {
+        $type: isRevertingTakedown
+          ? 'com.atproto.admin.defs#modEventReverseTakedown'
+          : 'com.atproto.admin.defs#modEventUnmute',
+        comment: comment ?? undefined,
+      },
+      createdAt,
+      createdBy,
+      subject,
+    })
+
+    if (isRevertingTakedown) {
+      if (subject.isRepo()) {
+        await this.reverseTakedownRepo(subject)
+      } else if (subject.isRecord()) {
+        await this.reverseTakedownRecord(subject)
+      }
+    }
+
+    return result
+  }
+
+  async takedownRepo(subject: RepoSubject, takedownId: number) {
+    await this.db.db
```

Comment:
we transactionally add the repo to the push_event table and then actually push it out & mark it as completed in the daemon

I'm considering attempting a push in a commit hook here, but not sure if it's worth the hassle. I kinda like the simplicity of the actual push living in just one place 🤔